### PR TITLE
Feature/4 overview tab

### LINKED ISSUE #4: Overview tab

### PR DESCRIPTION
**Pull request by timofei-maximov-saritasa**
#4

### CODE REVIEW

File: src/api/index.ts [hunk 1]
```diff
@@ -1,11 +1,20 @@
...
+    },
+  },
+});
 
 /**
  * TODO: CHANGE BASE URL.
  */
 export const http: AxiosInstance = axios.create({
-  baseURL: 'https://api.github.com/repos/tannerlinsley/react-query',
+  baseURL: 'https://vndbre-proxy.azurewebsites.net/api/v1/',
```
Comment (moetazaneta) [inline src/api/index.ts:10]: Move out to constants or env.

File: src/api/index.ts [hunk 2]
```diff
@@ -1,11 +1,20 @@
...
 import axios, { AxiosInstance } from 'axios';
 
-export const queryClient = new QueryClient();
+export const queryClient = new QueryClient({
+  defaultOptions: {
+    queries: {
+      retry: false,
+      refetchOnMount: false,
+      refetchOnWindowFocus: false,
+      refetchOnReconnect: false,
```
Comment (panov-aleksey-saritasa) [inline src/api/index.ts]: I don't think this setting should be global, it's better to make base config (and extend it if needed) and pass it into our custom useQuery hooks.

File: src/api/services/staffService.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ */
+export const fetchStaff = async(ids: number[]): Promise<Staff[]> => {
+  console.log(ids);
+  const { data } = await http.post<DataWrapper<StaffDto>>(
+    ApiUrls.VNDB,
+    `get staff basic (aid = [${ids}])`,
+  );
+
+  console.log(data);
+
```
Comment (panov-aleksey-saritasa) [inline src/api/services/staffService.ts]: Remove pls, you can check fetch results via react query devtools

File: src/features/VisualNovel/components/CharacterCard/CharacterCard.tsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import { Character } from '../../../../models/character';
+import cls from './CharacterCard.module.css';
+import { Roles } from '../../../../utils/types/roles';
+import { Icon } from '../../../../components/Icon/Icon';
+
+import characterPlaceholder from '../../../../assets/person.png';
+
+/**
+ * Component props.
+ */
```
Comment (panov-aleksey-saritasa) [inline src/features/VisualNovel/components/CharacterCard/CharacterCard.tsx]: Let's not write comments for props interface, only for props

File: src/features/VisualNovel/components/TagBlock/TagBlock.tsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  /**
+   * Block title.
+   */
+  readonly title: string;
+
+  /**
+   * Whether tag list is expandable or not.
+   */
+  readonly isExpandable?: boolean;
+
```
Comment (panov-aleksey-saritasa) [inline src/features/VisualNovel/components/TagBlock/TagBlock.tsx]: Extra line

File: src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx [hunk 6]
```diff
@@ -0,0 +1,210 @@
+import React, { FC, Fragment, useEffect, useState } from 'react';
+import { Heading, Link } from '@chakra-ui/react';
+
+import ISO6391 from 'iso-639-1';
```
Comment (panov-aleksey-saritasa) [inline src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx]: Maybe use alias for this module import?

File: src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+          tags={developers.map(dev => ({ name: dev }))}
+        />
+        {ISO6391.getAllCodes().map(key => (
+          <Fragment key={key}>
+            {publishers[key].length > 0 && (
+              <TagBlock
+                title={`Publisher (${ISO6391.getName(key)})`}
+                tags={publishers[key].map(publisher => ({ name: publisher }))}
+              />
+            )}
```
Comment (panov-aleksey-saritasa) [inline src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx]: Don't forget to refactor it 😄

File: src/utils/types/apiUrls.ts [hunk 8]
```diff
@@ -0,0 +1,8 @@
+export namespace ApiUrls {
+
+  /** Vndb proxy endpoint. */
+  export const VNDB = 'vndb';
+
+  /** Tags api endpoint. */
+  export const TAGS = 'tags';
```
Comment (panov-aleksey-saritasa) [inline src/utils/types/apiUrls.ts]: Why uppercase?

**Review by moetazaneta — CHANGES_REQUESTED** [review]

File: src/features/VisualNovel/pages/VisualNovelPage/VisualNovelPage.tsx [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+              <div className={cls.controls}>
+                <ButtonGroup isAttached>
+                  <Button mr="-px">Add to list</Button>
+                  <IconButton aria-label="Add to list extended" icon={<Icon name="carbon:chevron-down" />} />
+                </ButtonGroup>
+                <IconButton aria-label="Star" icon={<Icon name="carbon:star" />} colorScheme="gray" />
+                <IconButton aria-label="Edit" icon={<Icon name="carbon:edit" />} colorScheme="gray" />
+                <IconButton aria-label="Report" icon={<Icon name="carbon:flag" />} colorScheme="gray" />
+              </div>
+              <Text>{data.description}</Text>
```
Comment (moetazaneta) [inline src/features/VisualNovel/pages/VisualNovelPage/VisualNovelPage.tsx]: Add `white-space: break-spaces;` so it looks better 😉

File: src/api/index.ts (hunk 1 above)
Comment (prekel) [reply src/api/index.ts:10]: Will be done in #40

**Review by prekel — COMMENTED** [review]
lgtm

**Review by prekel — APPROVED** [review]
LGTM!!! 😉 🖖🖖🖖🖖

File: package.json [hunk 10]
```diff
@@ -3,13 +3,15 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "@chakra-ui/icons": "^1.0.16",
```
Comment (moetazaneta) [inline package.json]: Should be removed #16

File: src/api/dtos/characterDto.ts [hunk 11]
```diff
@@ -173,7 +173,7 @@ export interface CharacterDto {
    * "side",
    * "appears".
    */
-  vns: [number, number, number, string][];
+  vns?: [number, number, number, string][];
```
Comment (moetazaneta) [inline src/api/dtos/characterDto.ts:176]: I suggest move vn to separate interface, and properly doc it.

File: src/api/dtos/staffDto.ts [hunk 12]
```diff
@@ -100,20 +100,20 @@ export interface StaffDto {
    * the original name.
    * This list also includes the "primary" name.
    */
-  aliases: [number, string, string][];
+  aliases?: [number, string, string][];
```
Comment (moetazaneta) [inline src/api/dtos/staffDto.ts:103]: separate and properly doc

File: src/features/VisualNovel/pages/OverviewPage/OverviewPage.module.css [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+.link {
+  text-transform: capitalize;
+}
+
+.staff {
+  display: grid;
+  margin: 16px 0 40px 0;
+  grid-template-columns: repeat(3, 1fr);
+  grid-gap: 16px;
+
```
Comment (moetazaneta) [inline src/features/VisualNovel/pages/OverviewPage/OverviewPage.module.css]: extra line

File: src/api/dtos/staffDto.ts [hunk 14]
```diff
@@ -85,12 +85,12 @@ export interface StaffDto {
...
    * Links to staff's external resources.
    */
-  links: StaffLinks;
+  links?: StaffLinks;
 
   /**
    * Description or notes.
    */
-  description: string | null;
+  description?: string | null;
```
Comment (moetazaneta) [inline src/api/dtos/staffDto.ts]: Is it alright that type is `string | null | undefined`?

File: src/utils/types/apiUrls.ts (hunk 8 above)
Comment (moetazaneta) [reply src/utils/types/apiUrls.ts]: Why uppercase?

**Review by moetazaneta — CHANGES_REQUESTED** [review]

File: src/models/staff.ts [hunk 15]
```diff
@@ -106,30 +106,30 @@ export interface Staff {
...
    * Links to staff's external resources.
    */
-  links: StaffLinks;
+  links?: StaffLinks;
 
   /**
    * Description or notes.
    */
-  description: string | null;
+  description?: string | null;
```
Comment (moetazaneta) [inline src/models/staff.ts:114]: Again `null | undefined | string`

File: src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+    const publishersCopy = { ...publishers };
+    publisherReleases.forEach(release => {
+      release.languages.forEach(lang => {
+        const publisherNames = release.producers.map(p => p.name);
+        const uniquePublisherNames = Array.from(new Set(publishersCopy[lang].concat(publisherNames)));
+        publishersCopy[lang] = uniquePublisherNames;
+      });
+      setPublishers(publishersCopy);
+    });
+  };
```
Comment (panov-aleksey-saritasa) [inline src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx:55]: So `publishers` state is updated per each release? How does it work? Maybe just create result object after all iterations and call `setPublishers` one time?

File: src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+import { useVisualNovelQuery, useCharactersQuery, useReleasesQuery, useTagsQuery } from '../../queries';
+import { Release } from '../../../../models/release';
+
+/**
+ * Overview tab page.
+ */
+export const OverviewPage: FC = () => {
+  const { id } = useParams();
+  const [developers, setDevelopers] = useState<string[]>([]);
+  const [publishers, setPublishers] = useState<Record<string, string[]>>(
```
Comment (panov-aleksey-saritasa) [inline src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx:20]: Actually, as I see, `developers` and `publishers` are not state of the component, they are just dependable values that are calculated from `releases` data. Just create `const` variables from `fillDeveloperTeam` function. I think it should do the job.

File: src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+                      if (STAFF_ROLES[key as StaffRoles].showNote) {
+                        return { ...data, note: staff.note };
+                      }
+                      return data;
+                    })}
+                  />
+                )}
+              </Fragment>
+            ))
+          }
```
Comment (panov-aleksey-saritasa) [inline src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx:145]: Are you sure this piece of code should be in template?

**Review by panov-aleksey-saritasa — CHANGES_REQUESTED** [review]

File: src/features/VisualNovel/components/CharacterCard/CharacterCard.tsx [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+            <Text fontSize="xs">
+              Voice Actor
+            </Text>
+          </div>
+
+        </div>
+        {
+          role && (
+            <Text fontSize="xs">{Roles.toReadable(role)}</Text>
+          )
```
Comment (NightOfNights) [inline src/features/VisualNovel/components/CharacterCard/CharacterCard.tsx]: Why multiline?

File: src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx (hunk 17 above)
Comment (NightOfNights) [inline src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx]: Actually, as I see, `developers` and `publishers` are not state of the component, they are just dependable values that are calculated from `releases` data. Just create `const` variables from `fillDeveloperTeam` function. I think it should do the job.

File: src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx (hunk 16 above)
Comment (NightOfNights) [inline src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx]: So `publishers` state is updated per each release? How does it work? Maybe just create result object after all iterations and call `setPublishers` one time?

File: src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx (hunk 18 above)
Comment (NightOfNights) [inline src/features/VisualNovel/pages/OverviewPage/OverviewPage.tsx]: Are you sure this piece of code should be in template?

File: src/utils/types/staffRoles.ts [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+/** Object to display staff roles. */
+export const STAFF_ROLES: MappedStaffRoles = {
+  [StaffRoles.Director]: { title: 'Director' },
+  [StaffRoles.Scenario]: { title: 'Scenario' },
+  [StaffRoles.Artist]: { title: 'Artist' },
+  [StaffRoles.CharacterDesign]: { title: 'Character Designers' },
+  [StaffRoles.Music]: { title: 'Music', showNote: true },
+  [StaffRoles.Songs]: { title: 'Songs', showNote: true },
+  [StaffRoles.Staff]: { title: 'Staff', showNote: true },
+};
```
Comment (NightOfNights) [inline src/utils/types/staffRoles.ts]: Do you need `MappedStaffRoles` interface? Maybe just use `Record<StaffRoles, DisplayStaffRole>`?

**Review by NightOfNights — CHANGES_REQUESTED** [review]

File: src/api/mappers/visualNovelMapper.ts [hunk 21]
```diff
@@ -6,15 +6,38 @@ import {
...
+    case 3:
+      return VisualNovelLength.Medium;
+    case 4:
+      return VisualNovelLength.Long;
+    case 5:
+      return VisualNovelLength.VeryLong;
+    default:
+      return VisualNovelLength.Medium;
+  }
+};
```
Comment (NightOfNights) [inline src/api/mappers/visualNovelMapper.ts]: I think `VisualNovelLength` enum should be used for switch cases, not for returning of values

Comment (NightOfNights) [reply src/api/mappers/visualNovelMapper.ts]: Check how you did it in `src/utils/types/roles.ts` 😄

**Review by moetazaneta — APPROVED** [review]

**Review by NightOfNights — APPROVED** [review]

**Pull request merged**